### DISCUSSION TITLE
How to detect if an actor is on the ground?

### DISCUSSION THREAD

**Discussion by tenpaMk2** [excaliburjs/Excalibur#2406] [Q&A] [answered]
Hi, everyone. Please advise.

I want to detect if an actor is on the ground, because I want to make the player jumpable only when he is on the ground.
I have a problem to achieve it.

## sample code

- `index.ts`

```ts
Physics.gravity = Vector.Down.scale(300);
```

- `game-scene.ts`

```ts
import {
  Actor,
  CollisionType,
  Color,
  Engine,
  PreCollisionEvent,
  Scene,
  Vector,
} from "excalibur";

export class GameScene extends Scene {
  onInitialize(engine: Engine): void {
    const ground1 = new Actor({
      pos: new Vector(200, 700),
      width: 300,
      height: 200,
      color: Color.Cyan,
      collisionType: CollisionType.Fixed,
    });
    engine.add(ground1);

    const ground2 = new Actor({
      pos: new Vector(400, 400),
      width: 100,
      height: 600,
      color: Color.Magenta,
      collisionType: CollisionType.Fixed,
    });
    engine.add(ground2);

    const player = new Actor({
      x: engine.halfDrawWidth,
      y: engine.halfDrawHeight,
      width: 20,
      height: 20,
      color: Color.Chartreuse,
      collisionType: CollisionType.Active,
    });
    engine.add(player);

    player.vel = new Vector(30, 0);
    player.on("precollision", (event: PreCollisionEvent): void => {
      console.log(event.side);
    });
  }
}
```

## result

![スクリーンショット 2022-07-10 14 25 55](https://user-images.githubusercontent.com/8945311/178132515-8305469b-b444-4721-be57-7ad555794d73.png)

- `console.log`

```log
Bottom
Right
Bottom
Right
Bottom
Right
Bottom
Right
...
```

## problem

I get the log that repeats `Bottom` and `Right` .
It means 2 events are emitted each frame.

Therefore, I cannot detect if an actor is on the ground as following code.

```ts
    let isOnGround = false;
    player.on("precollision", (event: PreCollisionEvent): void => {
      if (event.side === Side.Bottom) {
        isOnGround = true;
      } else {
        isOnGround = false;
      }
    });
```

Does anyone knows good solution?

**Comment by tenpaMk2** [comment]
Here are all code.

<https://github.com/tenpaMk2/excalibur-examples/tree/main/examples/detect-ground>

**Comment by eonarheim** [comment, marked as answer]
@tenpaMk2 This is a good question!

There are a couple tricks that can be used to detect on ground in a situation like this:

1. You could use your most of your existing code and keep track of all sides collided in a frame, if `Side.Bottom` is in the list of sides that collided that frame you are on the ground.

  ```typescript
  export class GameScene extends Scene {
    playerOnGround = false;
    playerSideCollisions: Side[] = [];
    onPreUpdate() {
      playerSideCollisions.length = 0; // clear last frame's side collisions
    }
    ...
  
    isPlayerOnGround() {
      return playerSideCollisions.includes(Side.Bottom);
    }
  
    onInitialize() { 
      ...
      player.on("postcollision", (event: PostCollisionEvent): void => {
        playerSideCollisions.push(event.side);
      });
      engine.input.pointers.primary.on("down", (event: PointerEvent): void => {
        if (this.isPlayerOnGround()) player.vel = player.vel.add(new Vector(0, -100)); // jump!!
      });
    }
  ```
2. Use a child entity positioned just below the player and if it is currently colliding consider the player on the ground.

  ```typescript
  const playerGroup = ex.CollisionGroupManager.create('player'); // things in the same group do not collide
  const player = new Actor({
    x: engine.halfDrawWidth,
    y: engine.halfDrawHeight,
    width: 20,
    height: 20,
    color: Color.Chartreuse,
    collisionType: CollisionType.Active,
    collisionGroup: playerGroup
  });
  
  const groundColliderActor = new ex.Actor({
    x: 0,
    y: offsetBelowPlayer,
    width: playerWidth,
    height: 32,
    color: ex.Color.Green, // color there just to visualize where the collider is
    collisionType: ex.CollisionType.Passive,
    collisionGroup: playerGroup
  });
  
  let activeGroundCollisions = 0;
  groundColliderActor.on('collisionstart', () => {
    activeGroundCollisions++;
  });
  groundColliderActor.on('collisionend', () => {
    activeGroundCollisions--;
  });
  player.addChild(groundColliderActor);
  
  const isPlayerOnGround = () => {
    return activeGroundCollisions > 0;
  }
  ```

3. For fancier platforming, some people use ray casting to detect the floor but this can be a more complex approach. This has a couple advantages, mostly it's useful if you need more control over how your player interacts with the ground, like if you need to detect a slope, handle a character partially over an edge, or you need to know the distance to the ground. For Excalibur you currently [need a reference to every collider you want to raycast against](https://github.com/excaliburjs/Excalibur/discussions/2400#discussioncomment-3096634), which could get expensive or hard to use, this is something we want [to fix](https://github.com/excaliburjs/Excalibur/issues/2401) in the next version of Excalibur. 

```typescript
const floorActors: ex.Actor[] = [...];
const onGround = () => {
  for (let floor of floorActors) {
    const hit = floor.collider.get().rayCast(
      new ex.Ray(player.pos, ex.Vector.Down),
      100); // length to cast
    if (hit) {
      return true;
    }
    return false;
  }
}
```

**Reply by tenpaMk2** [reply]
@eonarheim Thank you for many solutions!!

## Solution 1

I love it. Very simple.

However, the player sometimes jumps too high.
I don't know why.

## Solution 2

Actually, I knew this because I use it in [dino-runner](https://github.com/tenpaMk2/excalibur-dino-runner) .

This also have a problem.
That is that the child emits the event when hit the right wall.
Therefore, the player can jump even though when not hit on the ground (bottom wall) .

I get around this problem by shortening the child.
However, this workaround occurs that incorrect detect...fmm...

## Solution 3

This is completely new idea for me!!
I can understand the concept by your code. Thanks.

## appendix

I implemented all solutions.

<https://github.com/tenpaMk2/excalibur-examples/tree/main/examples/detect-ground>
